onnxruntime/python/tools/quantization/fp16_converter.py

```diff
@@ -0,0 +1,359 @@
+        onnx_model = load_model('model.onnx')
+        new_onnx_model = convert_float_to_float16(onnx_model)
+        save_model(new_onnx_model, 'new_model.onnx')
+
+    """
+    func_infer_shape = None
+    if not disable_shape_infer and pv.Version(onnx.__version__) >= pv.Version("1.2"):
+        try:
+            from onnx.shape_inference import infer_shapes
+
+            func_infer_shape = infer_shapes
+        finally:
+            pass
+
+    if not isinstance(model, onnx_proto.ModelProto):
+        raise ValueError("Expected model type is an ONNX ModelProto but got %s" % type(model))
+
+    # create op_allow_list
+    if op_allow_list is None:
+        op_allow_list = ALLOWED_OPS_LIST
+    op_allow_list = set(op_allow_list)
+    # create a queue for BFS
+    queue = []
+    value_info_list = []
+    node_list = []
+    # type inference on input model
+    if func_infer_shape is not None:
+        model = func_infer_shape(model)
+    queue.append(model)
+    name_mapping = {}
+    graph_io_to_skip = set()
+    io_casts = set()
+    if keep_io_types:
+        for i, n in enumerate(model.graph.input):
+            if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                output_name = "graph_input_cast_" + str(i)
+                name_mapping[n.name] = output_name
+                graph_io_to_skip.add(n.name)
+
+                node_name = "graph_input_cast" + str(i)
+                new_value_info = model.graph.value_info.add()
+                new_value_info.CopyFrom(n)
+                new_value_info.name = output_name
+                new_value_info.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+                # add Cast node (from tensor(float) to tensor(float16) after graph input
+                new_node = [helper.make_node("Cast", [n.name], [output_name], to=10, name=node_name)]
+                model.graph.node.extend(new_node)
+                value_info_list.append(new_value_info)
+                io_casts.add(node_name)
+
+        for i, n in enumerate(model.graph.output):
+            if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                input_name = "graph_output_cast_" + str(i)
+                name_mapping[n.name] = input_name
+                graph_io_to_skip.add(n.name)
+
+                node_name = "graph_output_cast" + str(i)
+                # add Cast node (from tensor(float16) to tensor(float) before graph output
+                new_value_info = model.graph.value_info.add()
+                new_value_info.CopyFrom(n)
+                new_value_info.name = input_name
+                new_value_info.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+                new_node = [helper.make_node("Cast", [input_name], [n.name], to=1, name=node_name)]
+                model.graph.node.extend(new_node)
+                value_info_list.append(new_value_info)
+                io_casts.add(node_name)
+
+    while queue:
+        next_level = []
```

Comment:
is this a BFS? if it is, then why is 'next_level' needed?

Reply:
Removed this with queue.pop(0)